### PR TITLE
chrono privacy: no longer return `create` attribute in API

### PR DESCRIPTION
**Pull request by elrido**
This PR addresses a concern brought up in [#1290](https://github.com/orgs/PrivateBin/discussions/1290#discussioncomment-9244815)

## Changes
* stops adding a `create` attribute to the pastes metadata
* removes `create` or `postdate` attributes in pastes metadata, if they still exist
* drops `postdate` column in paste table

## ToDo
* [x] discuss: in the paste comments, the create date is internally used to sort the comments, as well as being displayed on the rendered comment section. Should we add an option to disable comment date display? If yes, we would still add and store it internally for sorting, but strip it before returning the comments array.
* [x] await merge of bootstrap 5 template, to ensure SRI gets updated in there as well

### CODE REVIEW

File: doc/Installation.md [hunk 1]
```diff
@@ -169,14 +169,13 @@ user these additional privileges:
...
 having to give PrivateBin too many permissions (replace `prefix_` with your own
-table prefix and create the table schema with your favourite MariaDB/MySQL
+table prefix and create the table schema with your favorite MariaDB/MySQL
 client):
 
 ```sql
 CREATE TABLE prefix_paste (
     dataid CHAR(16) NOT NULL,
     data MEDIUMBLOB,
-    postdate INT,
```
Comment (rugk) [inline doc/Installation.md:179]: Unlikely to break anything, but we have no (integration) tests that verify instances with the old DB schema still work with new PrivateBin?

**Review by rugk — CHANGES_REQUESTED** [review]
LGTM, but I don't see comments handling here? Likely because you did not change it – that still returns the creation time of the comments, does not it? **Edit:** ah, read it now.

**Comment by rugk** [comment]
> Should we add an option to disable comment date display? If yes, we would still add and store it internally for sorting, but strip it before returning the comments array.

Yeah, would be fine for me and would be an advantage.

Alternatively, for improved zero knowledge, we could make the DB field nullable and _never_ store it if the setting is disabled. The Frontend should then be able to handle non-timestamp entries there gracefully and just don't show a time then, i.e. if the setting is changed it could mean we have some comments shown with a timestamp and some without.
IMHO, that could increase privacy and is an acceptable and totally understandable (logical) tradeoff, when that setting is disabled.

**Review by rugk — APPROVED** [review]

**Comment by nain-F49FF806** [comment]
When enabling the option to not display comment dates, perhaps we should not even do any sorting?  (and thus not need to store the created date at all). 

Most database backends will likely return the comment entries in order they were added (even without a timestamp key for sorting).  Even if they don't, it's not a big deal.  

Even unsorted comments still have a tree structure because of their parent_id. So any conversation in the discussion will still be followable as comment threads.

**Comment by rugk** [comment]
> Most database backends will still likely return the comment entries in order they were added (even without a timestamp key for sorting). Even if they don't, it's not a big deal.

Well… the file backend likely does not guarantee this (but this is just my wild guess). But thanks yeah did not think of sorting when suggesting that we do not save the data at all.

One thing I'd like to avoid and disagree here is we should keep the order, always. Otherwise this can change meaning in the discussion when no tree-like structure was used. And as meaning is a content, we should mot introduce changes there… (Just imagine a comment "The previous comment is fake!" etc.) 
An alternative would be just to introduce (if not already there) a sequential ID (like a primary key in a DB) which you can use for sorting rather than a timestamp. Also, that should be faster/needs less CPU power…

**Comment by nain-F49FF806** [comment]
>  (Just imagine a comment "The previous comment is fake!" etc.)

I think such comments are best done as replies. Even on reddit / instagram or any popular platform, one does not rely on ordering of comments at the same level. Because comments are almost always reordered based on upvotes / algorithm or what not. The reply feature exists primarily for this use case, to refer explicitly to some previous comment.

**Comment by nain-F49FF806** [comment]
> An alternative would be just to introduce (if not already there) a sequential ID (like a primary key in a DB) which you can use for sorting rather than a timestamp. Also, that should be faster/needs less CPU power…

Yes, this would be good!. I thought of suggesting this too, but since this would not be applicable for file backends, it seemed an incomplete solution. 

I do feel that even without temporal order, comments have value. If the ordering is relevant, the option to display dates can still be left enabled.

File: doc/Installation.md (hunk 1 above)
Comment (elrido) [reply doc/Installation.md:179]: Yes we do. Hence I had to update a number of test cases as well. Thanks to these upgrade tests I discovered that not all SQLite versions support "DROP COLUMN" and added that version check (the version used with PHP in my unit test container and on Github do support it, but Scrutinizer uses an older one).
https://github.com/PrivateBin/PrivateBin/blob/ec02afca0469017a77fa5c8b952241c879883882/tst/Data/DatabaseTest.php#L326

We are limited by only testing SQLite but not on MariaDB, MySQL, Postgres, etc. If we wanted to test on those, we could extend the ControllerTest class, like we do for SQLite and keep those tests disabled by default, but for example on some dedicated Github actions, set up MariaDB & PG environments and enable these tests to run there.

**Comment by elrido** [comment]
So the problem that the unix timestamp solves for us, is that only the database backend stores the value separately, making it easy to sort. In all other backends, that value is stored inline and the order is not guaranteed[0]. That means to sort by it, all comments have to first get read, kept in memory and sorted there. Now, yes, in the database backend we could simply use the auto-incrementing primary key to sort by. But in the other backends (files & cloud storage) we would need to add a new "bucket" to track that serial number per paste and then hope we don't run into concurrency or locking issues if several comments occur at the same time.

In short, the current scheme is simple (and elegant?) in that it requires only storing and reading whole pastes and comments and no updates (only deletes of the whole things). Comments can be created without any knowledge of other comments. The unix timestamp happens to be roughly monotonic and for the cases it is not, the only side-effect is that some comments may get ordered incorrectly (i.e. during leap second smearing, hard NTP time shifts or if posted in the same second).

This is why I asked, as it could turn into a quite severe rewrite of the backend storage if we wanted to remove the timestamps from the comment storage, the way I did it for the pastes (where it really serves no purpose).

Since you all seem in favor to remove it from comments as well, I will add the configuration option for this ("display comment creation time", default true) and remove the internal value from being returned if that gets disabled, but for now will keep the internal time-based mechanism.

If you can think of any other value that isn't time (or at least not by itself), but available to PHP and at least roughly monotonically increasing, we could swap that in (kinda, existing comments would need to keep using time, but new pastes could be marked to allow for this new value). To work for all backends, it would need to be something that requires no extra storage and no synchronization between requests. It needed to be something inherent to the PHP runtime, that survives PHP restarts, server migrations, etc.

[0] for example in the filesystem backend, the comments are stored as files under a name made from paste, parent and comment IDs, and at best these are returned in lexical order (though some filesystems don't guarantee that either).

File: cfg/conf.sample.php [hunk 2]
```diff
@@ -18,6 +18,11 @@
 ; preselect the discussion feature, defaults to false
 opendiscussion = false
 
+; enable or disable the diplay of dates & times in the comments, defaults to true
+; note that internally the creation time will still get tracked in order to sort
+; the comments by creation time, but you can choose not to display them
```
Comment (rugk) [inline cfg/conf.sample.php:23]:
```suggestion
; Note that internally the creation time will still get tracked in order to sort
; the comments by creation time, but you can choose not to display them.
```

Let's make it a proper sentence.

File: lib/Model/Paste.php [hunk 3]
```diff
@@ -164,7 +163,17 @@ public function getComment($parentId, $commentId = '')
      */
     public function getComments()
     {
-        return $this->_store->readComments($this->getId());
+        if ($this->_conf->getKey('discussiondatedisplay')) {
+            return $this->_store->readComments($this->getId());
+        }
+        return array_map(function($comment) {
+            foreach (array('created', 'postdate') as $key) {
```
Comment (rugk) [inline lib/Model/Paste.php:170]: Do we have some test ensuring this new API feature? I.e. that the keys are properly deleted? (As that is really a thing one does not manually test/is cumbersome to test manually.)

Comment (elrido) [reply lib/Model/Paste.php:170]: nope, I tested manually

Comment (rugk) [reply lib/Model/Paste.php:170]: Maybe it would be a good idea to add that … was my saying… automated tests are always better.

Comment (elrido) [reply lib/Model/Paste.php:170]: Oki, I'll add a test for that code path, also in the interest of retaining a high level of code coverage.

Comment (elrido) [reply lib/Model/Paste.php:170]: Test added in a6058705a410b56fcc621e1fbfcbefe051d42881

Comment (elrido) [reply lib/Model/Paste.php:170]: We did actually have tests that validated that the paste creation (and postdate for v1 pastes) is present, which I had to change to detect that the field is now _not_ present. I changed those in the current location, which is in the ControllerTest, and therefore also executed in the inherited ControllerDatabaseTest and ...GcsTest variants: https://github.com/PrivateBin/PrivateBin/pull/1295/files#diff-8977b8e7d85969361448323730a50efb0c879b761a480352b86ad95c7ef932baL715

So we do now have tests that ensure that:
- paste creation is not present (for the file, database & Gcs backends)
- comment creation is normally present, but absent when the configuration has it set to false
- database upgrade will work error free (for sqlite version with and without drop column support)
- the new table schema works and pastes can still be read from it

**Review by rugk — APPROVED** [review]

**Pull request merged**